File: app/src/main/java/io/github/subhamtyagi/ocr/ocr/ImageTextReader.java
package io.github.subhamtyagi.ocr.ocr;

import android.graphics.Bitmap;

import com.googlecode.tesseract.android.TessBaseAPI;

import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import io.github.subhamtyagi.ocr.utils.Constants;
import io.github.subhamtyagi.ocr.utils.Language;


/**
 * This class convert the image to text and return the text on image
 */
public class ImageTextReader {

    public static final String TAG = "ImageTextReader";
    private boolean success;

    public boolean isSuccess() {
        return success;
    }

    public void setSuccess(boolean success) {
        this.success = success;
    }


    /**
     * TessBaseAPI instance
     */
    private static volatile TessBaseAPI api;
    //  private static volatile TesseractImageTextReader INSTANCE;

    /**
     * initialize and train the tesseract engine
     *
     * @param path      a path to training data
     * @param languages language code i.e. selected by user
     * @return the instance of this class for later use
     */
    public static ImageTextReader getInstance(String path, Set<Language> languages, int pageSegMode, Map<String, String> parameters, boolean isParameterSet, TessBaseAPI.ProgressNotifier progressNotifier) {
        try {
            ImageTextReader imageTextReader = new ImageTextReader();
            api = new TessBaseAPI(progressNotifier);
            boolean s = api.init(path, languages
                    .stream()
                    .map(Language::getCode)
                    .collect(Collectors.joining("+")));
            api.setPageSegMode(pageSegMode);
            imageTextReader.setSuccess(s);
            if (isParameterSet)
                for (Map.Entry<String, String> entry : parameters.entrySet()) {
                    String key = entry.getKey();
                    String value = entry.getValue();
                    if (!key.equals(Constants.KEY_OCR_PSM_MODE)) {
                        api.setVariable(key, value);
                    }
                }
            return imageTextReader;
        } catch (Exception e) {
            return null;
        }

    }

    /**
     * get the text from bitmap
     *
     * @param bitmap a image
     * @return text on image
     */
    public String getTextFromBitmap(Bitmap bitmap) {
        api.setImage(bitmap);
        String textOnImage;
        try {
            //textOnImage = api.getUTF8Text();
            textOnImage = api.getHOCRText(1);
        } catch (Exception e) {
            return "Scan Failed: WTF: Must be reported to developer!";
        }
        if (textOnImage.isEmpty()) {
            return "Scan Failed: Couldn't read the image\nProblem may be related to Tesseract or no Text on Image!";
        } else return textOnImage;

    }

    /**
     * stop the image TEXT reader
     */
    public void stop() {
        api.stop();
    }

    /**
     * find the confidence or
     *
     * @return confidence
     */
    public int getAccuracy() {
        return api.meanConfidence();
    }

    /**
     * Closes down tesseract and free up all memory.
     */
    public void tearDownEverything() {
        api.recycle();
    }

    /**
     * Frees up recognition results and any stored image data,
     */
    public void clearPreviousImage() {
        api.clear();
    }

}

File: core/src/main/java/io/github/subhamtyagi/ocr/engine/ImageTextReader.kt
package io.github.subhamtyagi.ocr.engine

import android.graphics.Bitmap
import android.text.Html
import android.util.Log
import com.googlecode.tesseract.android.TessBaseAPI
import io.github.subhamtyagi.ocr.data.model.Language

class ImageTextReader {
    var api: TessBaseAPI
    var success: Boolean = false

    constructor(
        path: String,
        languages: Set<Language>,
        pageSegMode: Int,
        ocrMode:Int,
        parameters: Map<String, String>,
        isParameterSet: Boolean,
        progressNotifier: TessBaseAPI.ProgressNotifier
    ) {
        api = TessBaseAPI(progressNotifier)
        val languageString = languages.joinToString("+") { it.code }
        Log.d("ImageTextReader", "Initializing Tesseract with languages: $languageString at path: $path")
        
        success = if (isParameterSet) {
            api.init(
                path,
                languageString,
                ocrMode,
                parameters
            )

        } else{
            api.init(
                path,
                languageString
            )
        }
        Log.d("ImageTextReader", "Tesseract initialization success: $success")
        api.setPageSegMode(pageSegMode)

    }

    val TAG="ImageTextReader"
    /**
     * Get the text from bitmap
     */
    fun getTextFromBitmap(bitmap: Bitmap): String {
        if (!success) return "OCR engine was not initialized successfully."
        if (bitmap.isRecycled) return "Provided bitmap is recycled."

        return try {
            Log.d(TAG, "getTextFromBitmap: get text called with bitmap $bitmap")
            api.setImage(bitmap)
            Log.d(TAG, "getTextFromBitmap: image set")
            val textOnImage = api.getHOCRText(1) ?: ""
            Log.d(TAG, "getTextFromBitmap: text gethocrtext called")
            val cleanText = Html.fromHtml(textOnImage).toString().trim { it <= ' ' };
            Log.d(TAG, "getTextFromBitmap: get text html")
            if (textOnImage.isEmpty()) {
                "Scan Failed: Couldn't read the image\nProblem may be related to Tesseract or no Text on Image!"
            } else {
                cleanText
            }
        } catch (e: Exception) {
            "Scan Failed: WTF: Must be reported to developer!\n ${e.stackTraceToString()}"
        }
    }

    /**
     * Stop the image text reader
     */
    fun stop() {
        api.stop()
    }

    /**
     * Get the mean confidence
     */
    fun getAccuracy(): Int {
        return api.meanConfidence()
    }

    /**
     * Closes down tesseract and free up all memory.
     */
    fun tearDownEverything() {
        api.recycle()
    }

    /**
     * Frees up recognition results and any stored image data.
     */
    fun clearPreviousImage() {
        api.clear()
    }
}
